src/core/internal/array/construction.d

```diff
@@ -90,6 +98,32 @@ Tarr _d_arrayctor(Tarr : T[], T)(return scope Tarr to, scope Tarr from) @trusted
     assert(arr1 == arr2);
 }
 
+@safe unittest
+{
+    // Test that copy constructor works
+    static int counter;
+    static struct S
+    {
+        int val;
+        this(int v)
+        {
+            val = v;
+        }
+        this(ref typeof(this) rhs)
+        {
+            val = rhs.val;
+            counter++;
+        }
+    }
+
+    S[4] arr1;
+    S[4] arr2 = [S(0), S(1), S(2), S(3)];
+    _d_arrayctor(arr1[], arr2[]);
```

Comment:
I was expecting that if I wrote
`S[4] arr1 = arr2;` this would call `_d_arrayctor`, but it seems it's not the case.

Can someone shed some light on this, please? What am I missing?

Reply:
Yeah, the problem is the compiler does not rewrite array assignments that have structs with copy constructors as elements to _d_arrayctor. Take a look at this code [1][2]. It checks if the array has struct elements with a postblit, but not sign of copy ctor. This is trivial to solve by just adding a check for cp ctor, however, I think the course of action here should be: merge this first, then merge the compiler change, then make a PR to update the test in this one.

[1] https://github.com/dlang/dmd/blob/master/src/dmd/e2ir.d#L2939
[2] https://github.com/dlang/dmd/blob/master/src/dmd/e2ir.d#L3239

Reply:
I had already looked at [1] and [2] before, but they are not called.
For the following code
```
@safe unittest
{
    static int counter;
    static struct S
    {
        int val;
        this(int v)
        {
            val = v;
        }
        this(this)
        {
            counter++;
            writefln("Here with %s\n", counter);
        }
    }

    S[4] arr1;
    S[4] arr2 = [S(0), S(1), S(2), S(3)];
```

Using the line
```
    arr1 = arr2;
```
won't call either [1] or [2].

Changing the code to
```
arr1[] = arr2;
```
will call [1] but it won't call `_d_arrayctor`

Changing the code to
```
    S[4] arr2 = [S(0), S(1), S(2), S(3)];
    S[4] arr1 = arr2;
```
will call [2] but again won't call `_d_arrayctor`

I must say that I am puzzled.

Also, I don't think that [this code](https://github.com/dlang/dmd/blob/master/src/dmd/e2ir.d#L2954) could call the [templated](https://github.com/dlang/druntime/blob/master/src/core/internal/array/construction.d#L24) version of `_d_arrayctor`. I was expecting it to call the non template version found [src.rt.arrayassign](https://github.com/dlang/druntime/blob/master/src/rt/arrayassign.d#L170) but it doesn't call this either.

Reply:
My point was that [1] and [2] should add checks for the copy constructor also otherwise the correct branches will not be taken in your unittest. I don'y actually understand what you mean by "it will call [1]". Do you mean the branch is taken?

Reply:
> My point was that [1] and [2] should add checks for the copy constructor also otherwise the correct branches will not be taken in your unittest.

Yes, I saw this. The `unittest` in the comment above uses postblit, not cpctor.

> I don'y actually understand what you mean by "it will call [1]". Do you mean the branch is taken?

Yes, I mean the branch is taken

Reply:
This was actually a big messup on my part. I din't realize I had to recompile `Phobos` to get the `druntime` changes visible into my local `dmd` build; the confusion was also amplified by the fact that I was "proof-checking" that I am linking with the correct `druntime` by having some `printf`s in druntime: the catch is that I had them in templated functions which, you got it, will get instantiated at compile time in user code

Reply:
So, to sum it up, the course of action going forward is to hook the call to `_d_arrayctor` in dmd, as @RazvanN7 pointed out.

Before doing so, we'll have to convert `_d_arrayassign`, `_d_arrayassign_l` and `_d_arrayassign_r` to into templates and then change dmd to call the templated versions of `_d_arrayassign` and the current templated version of `_d_arrayctor` and their `set` cousins `_d_arraysetctor` and `_d_arraysetassign`